File: app/src/main/java/com/example/plog/ui/exchange/MatchConfirmFragment.java
package com.example.plog.ui.exchange;

import com.example.plog.network.dto.MatchRecommendResponse;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;

import com.example.plog.R;
import com.example.plog.network.RetrofitClient;
import com.example.plog.network.api.ExchangeMatchApi;
import com.example.plog.network.api.ExchangeRoomApi;
import com.example.plog.network.dto.ExchangeMatchRequest;
import com.example.plog.network.dto.ExchangeMatchResponse;
import com.example.plog.network.dto.ExchangeRoomResponse;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.chip.Chip;

import java.util.ArrayList;
import java.util.List;

import android.util.Log;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MatchConfirmFragment extends Fragment {

    private Long pendingMatchId = null;
    private final Handler pollingHandler = new Handler(Looper.getMainLooper());
    private Runnable pollingRunnable;
    private List<MatchRecommendResponse> recommendList = new ArrayList<>();
    private int currentPartnerIndex = 0;

    private TextView tvMessage;
    private TextView tvSimilarity;
    private LinearLayout layoutButtons;
    private MaterialButton btnCancel;

    public MatchConfirmFragment() {
        super(R.layout.fragment_match_confirm);
    }

    private long getMyUserId() {
        return requireActivity()
                .getSharedPreferences("plog_prefs", Context.MODE_PRIVATE)
                .getLong("userId", -1L);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        tvMessage = view.findViewById(R.id.tvMatchMessage);
        TextView tvUserNickname = view.findViewById(R.id.tvUserNickname);
        tvSimilarity = view.findViewById(R.id.tvSimilarity);
        Chip chipTag1 = view.findViewById(R.id.chipTag1);
        Chip chipTag2 = view.findViewById(R.id.chipTag2);
        Chip chipTag3 = view.findViewById(R.id.chipTag3);
        LinearLayout layoutButtons = view.findViewById(R.id.layoutButtons);
        this.layoutButtons = layoutButtons;
        MaterialButton btnAccept = view.findViewById(R.id.btnAccept);
        MaterialButton btnReject = view.findViewById(R.id.btnReject);
        btnCancel = view.findViewById(R.id.btnCancel);

        if (getArguments() != null) {
            long matchId = getArguments().getLong("matchId", -1L);
            if (matchId != -1L) {
                pendingMatchId = matchId;
                showWaitingUI();

                ExchangeMatchApi api = RetrofitClient.getClient().create(ExchangeMatchApi.class);
                api.getMatch(matchId, getMyUserId()).enqueue(new Callback<ExchangeMatchResponse>() {
                    @Override
                    public void onResponse(Call<ExchangeMatchResponse> call, Response<ExchangeMatchResponse> response) {
                        if (!isAdded()) return;
                        if (response.isSuccessful() && response.body() != null) {
                            String nickname = response.body().getRequesterNickname();
                            if (nickname != null) tvUserNickname.setText(nickname);
                            List<String> categories = response.body().getTopCategories();
                            if (categories != null) {
                                chipTag1.setText(categories.size() > 0 ? "#" + categories.get(0) : "");
                                chipTag2.setText(categories.size() > 1 ? "#" + categories.get(1) : "");
                                chipTag3.setText(categories.size() > 2 ? "#" + categories.get(2) : "");
                            }
                        }
                    }
                    @Override
                    public void onFailure(Call<ExchangeMatchResponse> call, Throwable t) {
                        Log.e("MatchConfirm", "매칭 정보 조회 실패: " + t.getMessage());
                    }
                });

                startPolling(tvUserNickname);
                btnCancel.setOnClickListener(v -> cancelMatch());
                return;
            }

            recommendList = (ArrayList<MatchRecommendResponse>) getArguments().getSerializable("recommendList");
            if (recommendList != null && !recommendList.isEmpty()) {
                updatePartnerUI(tvUserNickname, chipTag1, chipTag2, chipTag3);
            }
        }

        btnAccept.setOnClickListener(v -> {
            MatchRecommendResponse partner = (recommendList != null && !recommendList.isEmpty())
                    ? recommendList.get(currentPartnerIndex) : null;
            Long targetUserId = partner != null ? partner.getUserId() : null;

            showWaitingUI();

            ExchangeMatchApi api = RetrofitClient.getClient().create(ExchangeMatchApi.class);
            api.createMatch(new ExchangeMatchRequest(getMyUserId(), targetUserId))
                    .enqueue(new Callback<ExchangeMatchResponse>() {
                        @Override
                        public void onResponse(Call<ExchangeMatchResponse> call, Response<ExchangeMatchResponse> response) {
                            if (!isAdded()) return;
                            if (response.isSuccessful() && response.body() != null) {
                                pendingMatchId = response.body().getId();
                                Toast.makeText(requireContext(), "매칭 신청 완료!", Toast.LENGTH_SHORT).show();
                                startPolling(tvUserNickname);
                                btnCancel.setOnClickListener(cv -> cancelMatch());
                            } else {
                                Toast.makeText(requireContext(), "이미 진행 중인 매칭이 있어요.", Toast.LENGTH_SHORT).show();
                                showMatchUI();
                            }
                        }
                        @Override
                        public void onFailure(Call<ExchangeMatchResponse> call, Throwable t) {
                            if (!isAdded()) return;
                            Toast.makeText(requireContext(), "서버 연결에 실패했어요.", Toast.LENGTH_SHORT).show();
                            showMatchUI();
                        }
                    });
        });

        btnReject.setOnClickListener(v -> {
            if (recommendList != null && recommendList.size() > 1) {
                currentPartnerIndex = (currentPartnerIndex + 1) % recommendList.size();
                updatePartnerUI(tvUserNickname, chipTag1, chipTag2, chipTag3);
            }
        });
    }

    private void showWaitingUI() {
        tvMessage.setText("매칭 중... 상대방의 수락을 기다리고 있어요.");
        tvSimilarity.setVisibility(View.GONE);
        layoutButtons.setVisibility(View.GONE);
        btnCancel.setVisibility(View.VISIBLE);
    }

    private void showMatchUI() {
        tvMessage.setText("이 사용자에게 교환일기를 신청할까요?");
        tvSimilarity.setVisibility(View.VISIBLE);
        layoutButtons.setVisibility(View.VISIBLE);
        btnCancel.setVisibility(View.GONE);
    }

    private void cancelMatch() {
        stopPolling();
        if (pendingMatchId != null) {
            ExchangeMatchApi api = RetrofitClient.getClient().create(ExchangeMatchApi.class);
            api.rejectMatch(pendingMatchId).enqueue(new Callback<Void>() {
                @Override
                public void onResponse(Call<Void> call, Response<Void> response) {
                    if (!isAdded()) return;
                    pendingMatchId = null;
                    if (recommendList != null && !recommendList.isEmpty()) {
                        showMatchUI();
                    } else {
                        NavHostFragment.findNavController(MatchConfirmFragment.this)
                                .navigate(R.id.matchingFragment);
                    }
                }
                @Override
                public void onFailure(Call<Void> call, Throwable t) {
                    if (!isAdded()) return;
                    pendingMatchId = null;
                    NavHostFragment.findNavController(MatchConfirmFragment.this)
                            .navigate(R.id.matchingFragment);
                }
            });
        }
    }

    private void startPolling(TextView tvUserNickname) {
        stopPolling();
        ExchangeMatchApi api = RetrofitClient.getClient().create(ExchangeMatchApi.class);
        pollingRunnable = new Runnable() {
            @Override
            public void run() {
                if (pendingMatchId == null || !isAdded()) return;
                api.getMatch(pendingMatchId, getMyUserId()).enqueue(new Callback<ExchangeMatchResponse>() {
                    @Override
                    public void onResponse(Call<ExchangeMatchResponse> call, Response<ExchangeMatchResponse> response) {
                        if (!isAdded()) return;
                        if (response.isSuccessful() && response.body() != null) {
                            String status = response.body().getStatus();
                            if ("MATCHED".equals(status)) {
                                stopPolling();
                                ExchangeRoomApi roomApi = RetrofitClient.getClient().create(ExchangeRoomApi.class);
                                roomApi.getActiveRoom(getMyUserId()).enqueue(new Callback<ExchangeRoomResponse>() {
                                    @Override
                                    public void onResponse(Call<ExchangeRoomResponse> call, Response<ExchangeRoomResponse> response) {
                                        if (!isAdded()) return;
                                        Bundle bundle = new Bundle();
                                        if (response.isSuccessful() && response.body() != null) {
                                            bundle.putLong("roomId", response.body().getId());
                                        }
                                        NavHostFragment.findNavController(MatchConfirmFragment.this)
                                                .navigate(R.id.matchedFragment, bundle);
                                    }
                                    @Override
                                    public void onFailure(Call<ExchangeRoomResponse> call, Throwable t) {
                                        if (!isAdded()) return;
                                        NavHostFragment.findNavController(MatchConfirmFragment.this)
                                                .navigate(R.id.matchedFragment, new Bundle());
                                    }
                                });
                            } else if ("REJECTED".equals(status)) {
                                stopPolling();
                                Toast.makeText(requireContext(), "매칭이 거절됐어요.", Toast.LENGTH_SHORT).show();
                                NavHostFragment.findNavController(MatchConfirmFragment.this)
                                        .navigate(R.id.matchingFragment);
                            } else {
                                pollingHandler.postDelayed(pollingRunnable, 3000);
                            }
                        }
                    }
                    @Override
                    public void onFailure(Call<ExchangeMatchResponse> call, Throwable t) {
                        if (isAdded()) pollingHandler.postDelayed(pollingRunnable, 3000);
                    }
                });
            }
        };
        pollingHandler.post(pollingRunnable);
    }

    private void stopPolling() {
        if (pollingRunnable != null) {
            pollingHandler.removeCallbacks(pollingRunnable);
            pollingRunnable = null;
        }
    }

    private void updatePartnerUI(TextView tvNickname, Chip chip1, Chip chip2, Chip chip3) {
        if (recommendList == null || recommendList.isEmpty()) return;
        MatchRecommendResponse partner = recommendList.get(currentPartnerIndex);
        tvNickname.setText(partner.getNickname());
        List<String> categories = partner.getTopCategories();
        chip1.setText(categories != null && categories.size() > 0 ? "#" + categories.get(0) : "");
        chip2.setText(categories != null && categories.size() > 1 ? "#" + categories.get(1) : "");
        chip3.setText(categories != null && categories.size() > 2 ? "#" + categories.get(2) : "");
        TextView tvSimilarity = requireView().findViewById(R.id.tvSimilarity);
        tvSimilarity.setText("나와 " + (int)(partner.getSimilarityScore() * 100) + "% 일치해요!");
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        stopPolling();
    }
}

File: app/src/main/java/com/example/plog/ui/exchange/MatchConfirmFragment.kt
package com.example.plog.ui.exchange

import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.findNavController
import com.example.plog.R
import com.example.plog.MainActivity
import com.google.android.material.button.MaterialButton
import com.google.android.material.chip.Chip
import kotlin.random.Random

/**
 * 교환일기 신청 전 상대방의 프로필을 확인하고 매칭을 확정하는 프래그먼트
 */
class MatchConfirmFragment : Fragment(R.layout.fragment_match_confirm) {

    // ==========================================
    // 시연용 더미 데이터 세트
    // TODO: 추후 유저 성향 분석 서버 API 연동 필요
    // ==========================================
    private val partnerPool = listOf(
        listOf("user1", "#영화", "#드라마", "#집순이"),
        listOf("user2", "#카페", "#사진", "#여행"),
        listOf("user3", "#개발", "#독서", "#고양이"),
        listOf("user4", "#맛집", "#베이킹", "#디저트"),
        listOf("user5", "#음악", "#기타", "#감성")
    )

    private var currentPartnerIndex = 0

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // UI 컴포넌트 바인딩
        val tvMessage = view.findViewById<TextView>(R.id.tvMatchMessage)
        val tvUserNickname = view.findViewById<TextView>(R.id.tvUserNickname)

        val chipTag1 = view.findViewById<Chip>(R.id.chipTag1)
        val chipTag2 = view.findViewById<Chip>(R.id.chipTag2)
        val chipTag3 = view.findViewById<Chip>(R.id.chipTag3)

        val layoutButtons = view.findViewById<LinearLayout>(R.id.layoutButtons)
        val btnAccept = view.findViewById<MaterialButton>(R.id.btnAccept)
        val btnReject = view.findViewById<MaterialButton>(R.id.btnReject)
        val btnCancel = view.findViewById<MaterialButton>(R.id.btnCancel)

        // 초기 가상 사용자 데이터 로드
        setNewPartner(tvUserNickname, chipTag1, chipTag2, chipTag3)

        // ==========================================
        // 버튼 이벤트 리스너 설정
        // ==========================================

        // 교환 신청 수락 ("예" 버튼 클릭)
        btnAccept.setOnClickListener {
            tvMessage.text = "상대방의 응답을 기다리는 중..."

            layoutButtons.visibility = View.GONE
            btnCancel.visibility = View.VISIBLE

            // 비동기 매칭 수락 시뮬레이션 (2초 대기)
            view.postDelayed({
                // 시스템 상단 노티피케이션 알림 발생
                (activity as? MainActivity)?.showNotification(
                    "매칭 성공!",
                    "${tvUserNickname.text}님이 신청을 수락했습니다."
                )

                // 알림 노출 후 최종 일기장 화면으로 이동 (1.5초 대기)
                view.postDelayed({
                    val bundle = Bundle().apply {
                        putString("partnerName", tvUserNickname.text.toString())
                    }
                    findNavController().navigate(
                        R.id.matchedFragment,
                        bundle
                    )
                }, 1500)

            }, 2000)
        }

        // 재탐색 ("다시 찾기" 버튼 클릭)
        btnReject.setOnClickListener {
            // 중복 방지 처리가 적용된 렌덤 인덱스 추출
            var nextIndex = currentPartnerIndex
            if (partnerPool.size > 1) {
                while (nextIndex == currentPartnerIndex) {
                    nextIndex = Random.nextInt(partnerPool.size)
                }
            } else {
                nextIndex = 0
            }
            currentPartnerIndex = nextIndex

            // 새로운 유저 프로필 갱신
            setNewPartner(tvUserNickname, chipTag1, chipTag2, chipTag3)
        }

        // 신청 취소 버튼 클릭
        btnCancel.setOnClickListener {
            tvMessage.text = "이 사용자에게 교환일기를 신청할까요?"

            layoutButtons.visibility = View.VISIBLE
            btnCancel.visibility = View.GONE
        }
    }

    // ==========================================
    // UI 데이터 바인딩 헬퍼 함수
    // ==========================================

    /**
     * 지정된 인덱스의 가상 유저 데이터(닉네임 및 태그)를 뷰에 매핑합니다.
     */
    private fun setNewPartner(tvNickname: TextView, chip1: Chip, chip2: Chip, chip3: Chip) {
        val partnerData = partnerPool[currentPartnerIndex]

        tvNickname.text = partnerData[0]
        chip1.text = partnerData[1]
        chip2.text = partnerData[2]
        chip3.text = partnerData[3]
    }
}